File: filly-core/src/widgets/checklist.rs
use std::collections::HashSet;
use crate::event::{Event, KeyCode};
use crate::render::{
    Alignment, BorderStyle, EdgeInsets, ListItem, Rect, RenderNode, RenderTree, TextStyle,
};
use crate::widget::{EventResult, Widget};
use filly_protocol::WidgetResponse;

pub struct ChecklistWidget {
    pub title: String,
    pub message: String,
    pub choices: Vec<String>,
    pub checked: HashSet<usize>,
    pub selected: usize,
    pub min: usize,
    pub max: usize,
    dirty: bool,
}

impl ChecklistWidget {
    pub fn new(
        title: String,
        message: String,
        choices: Vec<String>,
        min: Option<usize>,
        max: Option<usize>,
        default: Vec<String>,
    ) -> Self {
        let checked: HashSet<usize> = choices
            .iter()
            .enumerate()
            .filter(|(_, c)| default.contains(c))
            .map(|(i, _)| i)
            .collect();
        Self {
            title,
            message,
            choices,
            checked,
            selected: 0,
            min: min.unwrap_or(0),
            max: max.unwrap_or(usize::MAX),
            dirty: true,
        }
    }
}

impl Widget for ChecklistWidget {
    fn render(&self, area: Rect) -> RenderTree {
        let box_w = ((area.width as f32 * 0.7) as u16).min(area.width.saturating_sub(2));
        let box_h = ((area.height as f32 * 0.8) as u16).min(area.height.saturating_sub(2));
        let box_x = (area.width.saturating_sub(box_w)) / 2;
        let box_y = (area.height.saturating_sub(box_h)) / 2;

        let mut children = Vec::new();
        if !self.title.is_empty() {
            children.push(RenderNode::Text {
                rect: Rect::new(1, 0, box_w - 2, 1),
                content: self.title.clone(),
                alignment: Alignment::Center,
                role: None, label: None, description: None, style: TextStyle { bold: true, fg: Some("green".into()), ..TextStyle::normal() },
            });
        }
        if !self.message.is_empty() {
            children.push(RenderNode::Text {
                rect: Rect::new(1, 1, box_w - 2, 2),
                content: self.message.clone(),
                alignment: Alignment::Left,
                style: TextStyle::normal(), role: None, label: None, description: None,
            });
        }

        let list_y = if self.message.is_empty() { 1 } else { 3 };
        // List takes everything from list_y down to 2 rows above the bottom
        let list_h = box_h.saturating_sub(list_y).saturating_sub(2);

        let items: Vec<ListItem> = self.choices.iter().enumerate().map(|(i, c)| {
            let mark = if self.checked.contains(&i) { "[x]" } else { "[ ]" };
            ListItem { label: format!(" {} {}", mark, c), meta: None }
        }).collect();

        let selected = self.selected.min(self.choices.len().saturating_sub(1));
        children.push(RenderNode::List {
            rect: Rect::new(1, list_y, box_w - 2, list_h),
            items,
            selected: Some(selected),
            highlight_style: TextStyle::selected(), role: None, label: None, description: None,
        });

        // Footer lines pinned to bottom
        children.push(RenderNode::Text {
            rect: Rect::new(1, box_h - 2, box_w - 2, 1),
            content: format!(" Selected: {}/{}   Space=toggle  Enter=confirm  Esc=cancel", self.checked.len(), self.choices.len()),
            alignment: Alignment::Left,
            style: TextStyle::muted(), role: None, label: None, description: None,
        });
        children.push(RenderNode::Text {
            rect: Rect::new(1, box_h - 1, box_w - 2, 1),
            content: "Up/Down:move  Space:toggle  Enter:confirm  Esc:cancel".into(),
            alignment: Alignment::Center,
            style: TextStyle::muted(), role: None, label: None, description: None,
        });

        RenderTree::Container {
            rect: Rect::new(box_x, box_y, box_w, box_h),
            background: None,
            border: BorderStyle::Single,
            padding: EdgeInsets::zero(), role: None, label: None, description: None,
            children,
        }
    }

    fn handle_event(&mut self, event: &Event) -> EventResult {
        match event {
            Event::Key(code, _) => match code {
                KeyCode::Esc => EventResult::Response(WidgetResponse {
                    result: None,
                    cancelled: true,
                    error: None,
                }),
                KeyCode::Up => {
                    if self.selected > 0 {
                        self.selected -= 1;
                        self.dirty = true;
                    }
                    EventResult::Handled
                }
                KeyCode::Down => {
                    if self.selected + 1 < self.choices.len() {
                        self.selected += 1;
                        self.dirty = true;
                    }
                    EventResult::Handled
                }
                KeyCode::Char(' ') => {
                    if self.checked.contains(&self.selected) {
                        if self.checked.len() > self.min {
                            self.checked.remove(&self.selected);
                        }
                    } else if self.checked.len() < self.max {
                        self.checked.insert(self.selected);
                    }
                    self.dirty = true;
                    EventResult::Handled
                }
                KeyCode::Enter => {
                    if self.checked.len() >= self.min && self.checked.len() <= self.max {
                        let selected: Vec<String> = self
                            .checked
                            .iter()
                            .map(|&i| self.choices[i].clone())
                            .collect();
                        EventResult::Response(WidgetResponse {
                            result: Some(serde_json::Value::Array(
                                selected
                                    .into_iter()
                                    .map(serde_json::Value::String)
                                    .collect(),
                            )),
                            cancelled: false,
                            error: None,
                        })
                    } else {
                        EventResult::Handled
                    }
                }
                _ => EventResult::Unhandled,
            },
            _ => EventResult::Unhandled,
        }
    }

    fn is_dirty(&self) -> bool {
        self.dirty
    }
    fn clear_dirty(&mut self) {
        self.dirty = false;
    }
}

File: filly-core/widgets/checklist.c
#include <stdio.h>
#include "checklist.h"
#include <stdlib.h>
#include <string.h>

typedef struct {
    char *title;
    char *message;
    char **choices;
    int count;
    bool *checked;
    int selected;
    int min_sel;
    int max_sel;
    bool dirty;
} ChecklistData;

static void checklist_render(Widget *self, Rect area, RenderTree *out) {
    ChecklistData *d = (ChecklistData *)(self + 1);
    memset(out, 0, sizeof(*out));
    int box_w = (int)(area.w * 0.7f);
    if (box_w > area.w - 2) box_w = area.w - 2;
    int box_h = (int)(area.h * 0.8f);
    if (box_h > area.h - 2) box_h = area.h - 2;
    int box_x = (area.w - box_w) / 2, box_y = (area.h - box_h) / 2;

    RenderTree *children = calloc(4 + 1, sizeof(RenderTree));
    int idx = 0;
    if (d->title && strlen(d->title)) {
        RenderTree *t = &children[idx++];
        t->type = RNODE_TEXT;
        t->rect = rect_new(1, 0, box_w - 2, 1);
        t->text.content = strdup(d->title);
        t->text.align = ALIGN_CENTER;
        t->text.style = textstyle_selected();
    }
    int list_y = 1;
    if (d->message && strlen(d->message)) {
        RenderTree *t = &children[idx++];
        t->type = RNODE_TEXT;
        t->rect = rect_new(1, 1, box_w - 2, 2);
        t->text.content = strdup(d->message);
        t->text.align = ALIGN_LEFT;
        t->text.style = textstyle_normal();
        list_y = 3;
    }
    int list_h = box_h - list_y - 2;
    RenderTree *list = &children[idx++];
    list->type = RNODE_LIST;
    list->rect = rect_new(1, list_y, box_w - 2, list_h > 0 ? list_h : 1);
    list->list.item_count = d->count;
    list->list.items = malloc(d->count * sizeof(ListItem));
    for (int i = 0; i < d->count; i++) {
        char label[1024];
        snprintf(label, sizeof(label), " %s %s", d->checked[i] ? "[x]" : "[ ]", d->choices[i]);
        list->list.items[i] = listitem_new(label);
    }
    list->list.selected = d->selected;
    list->list.highlight = textstyle_selected();

    RenderTree *footer = &children[idx++];
    footer->type = RNODE_TEXT;
    footer->rect = rect_new(1, box_h - 2, box_w - 2, 1);
    int checked_count = 0;
    for (int i = 0; i < d->count; i++) if (d->checked[i]) checked_count++;
    char footer_text[256];
    snprintf(footer_text, sizeof(footer_text), "Selected: %d/%d  Space:toggle  Enter:confirm  Esc:cancel", checked_count, d->count);
    footer->text.content = strdup(footer_text);
    footer->text.align = ALIGN_CENTER;
    footer->text.style = textstyle_muted();

    out->type = RNODE_CONTAINER;
    out->rect = rect_new(box_x, box_y, box_w, box_h);
    out->container.border = BORDER_SINGLE;
    out->container.padding = edgeinsets_zero();
    out->container.children = children;
    out->container.child_count = idx;
}

static EventResult checklist_handle_event(Widget *self, Event *ev, Backend *backend) {
    ChecklistData *d = (ChecklistData *)(self + 1);
    if (ev->type != EVENT_KEY) return event_result_unhandled();
    switch (ev->code) {
        case KEY_ESC:
            return event_result_response((WidgetResponse){ .result = NULL, .cancelled = true, .error = NULL });
        case KEY_UP:
            if (d->selected > 0) { d->selected--; d->dirty = true; }
            return event_result_handled();
        case KEY_DOWN:
            if (d->selected + 1 < d->count) { d->selected++; d->dirty = true; }
            return event_result_handled();
        case KEY_CHAR:
            if (ev->ch == ' ') {
                int checked_count = 0;
                for (int i = 0; i < d->count; i++) if (d->checked[i]) checked_count++;
                if (d->checked[d->selected]) {
                    if (checked_count > d->min_sel) { d->checked[d->selected] = false; d->dirty = true; }
                } else {
                    if (checked_count < d->max_sel) { d->checked[d->selected] = true; d->dirty = true; }
                }
            }
            return event_result_handled();
        case KEY_ENTER: {
            int checked_count = 0;
            for (int i = 0; i < d->count; i++) if (d->checked[i]) checked_count++;
            if (checked_count >= d->min_sel && checked_count <= d->max_sel) {
                cJSON *arr = cJSON_CreateArray();
                for (int i = 0; i < d->count; i++)
                    if (d->checked[i]) cJSON_AddItemToArray(arr, cJSON_CreateString(d->choices[i]));
                return event_result_response((WidgetResponse){ .result = arr, .cancelled = false, .error = NULL });
            }
            return event_result_handled();
        }
        default:
            return event_result_unhandled();
    }
}

static bool checklist_is_dirty(Widget *self) { return ((ChecklistData *)(self + 1))->dirty; }
static void checklist_clear_dirty(Widget *self) { ((ChecklistData *)(self + 1))->dirty = false; }
static void checklist_destroy(Widget *self) {
    ChecklistData *d = (ChecklistData *)(self + 1);
    free(d->title); free(d->message);
    for (int i = 0; i < d->count; i++) free(d->choices[i]);
    free(d->choices); free(d->checked);
}

Widget *checklist_widget_new(const char *title, const char *message, char **choices, int count,
                              int min_sel, int max_sel, char **default_choices, int default_count) {
    Widget *w = calloc(1, sizeof(Widget) + sizeof(ChecklistData));
    w->vtable.render = checklist_render;
    w->vtable.handle_event = checklist_handle_event;
    w->vtable.is_dirty = checklist_is_dirty;
    w->vtable.clear_dirty = checklist_clear_dirty;
    w->vtable.destroy = checklist_destroy;
    ChecklistData *d = (ChecklistData *)(w + 1);
    d->title = title ? strdup(title) : NULL;
    d->message = message ? strdup(message) : NULL;
    d->choices = malloc(count * sizeof(char *));
    for (int i = 0; i < count; i++) d->choices[i] = strdup(choices[i]);
    d->count = count;
    d->checked = calloc(count, sizeof(bool));
    if (default_choices)
        for (int i = 0; i < default_count; i++)
            for (int j = 0; j < count; j++)
                if (strcmp(d->choices[j], default_choices[i]) == 0) d->checked[j] = true;
    d->selected = 0;
    d->min_sel = min_sel;
    d->max_sel = max_sel > 0 ? max_sel : count;
    d->dirty = true;
    return w;
}